### DISCUSSION TITLE
Edge Server Actions

### DISCUSSION THREAD

**Discussion by kiikoh** [vercel/next.js#49605] [Help] [answered]
### Summary

How can Edge Functions be configured in server actions?
```tsx
"use server";

import { revalidatePath } from "next/cache";
import { db } from "~/db";
import { presses } from "~/db/schema";

export const increment = async () => {
  console.log("yo");
  await db.insert(presses).values({});
  revalidatePath("/");
};

export const config = {
  runtime: "edge",
  regions: ["iad1"],
};
```
I tried this but I get the following error in vercel deployment... 
```
Error: 
  x Only async functions are allowed to be exported in a "use server" file.
    ,-[/vercel/path0/src/app/action.ts:10:1]
 10 |       revalidatePath("/");
 11 |     };
 12 |     
 13 | ,-> export const config = {
 14 | |     runtime: "edge",
 15 | |     regions: ["iad1"],
 16 | `-> };
    `----

```

**Comment by remorses** [comment]
try `export const runtime = 'edge'`

**Reply by kiikoh** [reply]
Same error

**Reply by okdonga** [reply]
Doesn't work for me either.

**Comment by okdonga** [comment]
i'm also looking for ways to make server actions into edge functions due to a rate limit of 10 sec with serverless functions...  would love to know whether this is possible at this stage.. since server action just got released.. and if so how to configure! 

`export const config = {
  runtime: "edge",
  regions: ["iad1"],
};`

for me, with the above config setting, vercel deploys successfully. but the config doesn't seem to make any changes. it's still the same serverless function.

**Comment by okdonga** [comment]
Hi [kiikoh](https://github.com/kiikoh), I kind of made a workaround and made it work. 

For me, there was just no error, but the function just didn't register as an edge function on vercel.
So, instead of using server actions, I ended up adding a route handler, and made the front-end call this route handler. 
Again, just adding the config in the server actions didn't register the function as an edge function for me. 

In case you want to convert to using a route handler, here are some troublesshooting I did along the way: 

1. I first created an API and made sure the front-end could successfully call it.

2. Then, added the following code to the route file to register as edge functions: 
```
export const runtime = 'edge';
export const preferredRegion = 'iad1'; 
```

The old configuration syntax (using `export const config`) resulted in the following error. 

```
Error: `export const config` in /vercel/path0/src/app/api/openai/route.ts is deprecated. 
Please change `runtime` property to segment export config. 
See https://beta.nextjs.org/docs/api-reference/segment-config
```
So, I followed the guide provided [here](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config).

If I were you, I would first try removing the `'use server'` keyword in your action file, as that seems to be the issue that next.js is flagging. 
But, even then, I was unable to register server action as edge functions..  please share if you find any other ways!

**Reply by kiikoh** [reply]
Ok, I'm assuming its just an "alpha" thing... hopefully they can get that fixed at some point

**Comment by weleo** [comment]
Do server actions run by default on node?

**Comment by jadedevin13** [comment]
So I was able to make it work on my end.

<img width="867" alt="image" src="https://github.com/vercel/next.js/assets/6671182/fb50c841-907f-457a-97c5-76066e41d26b">

Do **NOT** use "use server" on top of the file. Use "use server" on every action instead. Do **NOT** use `import { PageConfig } from "next";`. export the runtime as a string and not part of an object.


Like this.
```
export const runtime = "edge";

export async function getData(data: any) {
  "use server"

  console.log(data)
  return {'wew': 'wew'}
}
```

**Reply by okdonga** [reply]
@jadedevin13 Hi, I tried your method as suggested.. but still runtime is set as node.js 
Example code on the [Vercel's official documentation ](https://vercel.com/docs/concepts/functions/edge-functions/quickstart) shows to export runtime as part of an object though. 
```
import { NextRequest, NextResponse } from 'next/server';
 
export const config = {
  runtime: 'edge',
};
 
export default (request: NextRequest) => {
  return NextResponse.json({
    name: `Hello, from ${request.url} I'm now an Edge Function!`,
  });
};
```
https://vercel.com/docs/concepts/functions/edge-functions/quickstart

**Comment by Lermatroid** [comment]
Any updates on this? Still can't find any docs that give a solid answer.

**Comment by Xanacas** [comment]
I'm also looking for a way to deal with this

**Reply by arthureberledev** [reply, marked as answer]
Maybe this helps you:

> Server Actions inherit the [runtime](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes) from the page or layout they are used on.

[Source](https://nextjs.org/docs/app/building-your-application/data-fetching/forms-and-mutations#how-server-actions-work)

**Reply by kiikoh** [reply]
Ah, that is interesting. Makes sense but doesn't match the same patterns as other routes. I feel like they shouldn't be tied together... I can imagine a scenario where the data fetching on a page should be edge but maybe an action needs a node runtime.

**Reply by Lermatroid** [reply]
Is this the case even if the action is in a separate file (eg: `action.ts`)?

**Reply by soymartinez** [reply]
even if I use the action from the client side? 🤔

**Reply by ajayvignesh01** [reply]
Came here to ask the same question lol

**Reply by HarshaSurampudi** [reply]
I have this exact need. I need to make a db call (which can only be done in node runtime) inside an edge runtime endpoint. If I use server action, it is inheriting the edge runtime and the db call cannot happen. As a workaround I am making an api call to another node runtime endpoint. Any ideas for a better solution? Thanks.

**Comment by sannajammeh** [comment]
@leerob Is there any reason why server actions can't have their own runtime? Take this example:

A `/products` page fetches all products once - Cached via SSG. 
The page calls `getProducts({...filter})` which is a server action.

Once fetched products are moved into a client component for other data fetching like search queries, filtering etc. This client component also uses the same `getProducts` server action. 

- We don't want to run this server action on Node as it can have a slow startup time.
- We use a server action to leverage Next.js fetch caching easily while having good DX.
- If we use route handlers we must type this manually.

Now for the problem:
If we set the whole runtime to edge on `/products` we lose SSG. If we leave it on Node.js we keep SSG and all functionality but get a 300ms startup time penalty.

**Reply by TonsiTT** [reply]
Did you ever find a solution?

**Reply by rafalzawadzki** [reply]
looks like the only workaround for now is implementing a dedicated API route... bummer

**Comment by watjurk** [comment]
Hello!
The ability to set the runtime on the server action itself would be very useful for me, are there any plans to add it in the future? (if not consider this a feature request).

My specific use-case is a nextjs app deployed using cloudflare's [next-on-pages](https://github.com/cloudflare/next-on-pages) adapter. For server actions to work on cloudflare the runtime must be set to edge. Unfortunately, for now, the only possibility to achieve that is to select runtime to edge on the Page calling the server action, which is not ideal.

Best

**Comment by leerob** [comment]
We are not planning to make further improvements to the edge runtime. Instead, we are working to expand usage of the Node.js runtime (including support for using Middleware with Node.js).
